_posts/2019-08-23-unnest-operator-performance-enhancements.md

```diff
@@ -0,0 +1,170 @@
+---
+layout: post
+title:  "Unnest Operator Performance Enhancement with Dictionary Blocks"
+author: Pratham Desai, LinkedIn
+excerpt_separator: <!--more-->
+---
+
+Queries with `CROSS JOIN UNNEST` clause are expected to have a significant performance improvement starting version 316.
+
+<!--more-->
+
+# Executive Summary
+
+The execution plans for queries with a `CROSS JOIN UNNEST` clause contain an Unnest Operator. The previous implementation of Unnest Operator performed a deep copy on all input blocks to generate output blocks. This caused high CPU consumption and memory allocation for the operator, and impacted the performance of such queries. The impact was worse for `UNNEST` queries accessing a high number of columns, or even a few columns with deeply nested schema.
+
+<!--more-->
+
+We realized that the implementation can be made more efficient by avoiding copies in the Unnest Operator, if possible. Using dictionary blocks to create output blocks pointing to input elements has given us significant CPU and memory benefits by avoiding copies. The benchmark results for the new Unnest Operator implementation show more than ~10x gain in CPU time and 3x~5x gain in memory allocation.
+
+<!--more-->
+
+Let’s try to understand this change with an example. At LinkedIn, the most common usage for `CROSS JOIN UNNEST` clause is seen to be for unnesting a single array or map column. A sample query with the clause would look like the following:
+
+```sql
+SELECT T.c0, U.unnest_c1 
+FROM T CROSS JOIN UNNEST(c1) AS U(unnest_c1)
+```
+
+The plots below compare the performance of Unnest Operator in the previous and the current implementation for 3 different cases. Every case evaluates the Unnest Operator performance for a query like the above, on a table `T` with two columns `c0` and `c1`. For all the 3 cases, `c0` is a `VARCHAR` type column. But the nested column `c1` is of `ARRAY<VARCHAR>`, `MAP<VARCHAR, VARCHAR>` and `ARRAY<ROW(VARCHAR, VARCHAR, VARCHAR)>` types respectively. All the `VARCHAR` elements in both the columns have length 50, and the arrays in the second column have lengths distributed uniformly between 0 and 300. 
```

Comment:
The canonical form for parametric types is `ARRAY(VARCHAR)`, `MAP(VARCHAR, VARCHAR)`, etc. The angle brackets are a legacy form.

Reply:
I see. Fixed.